src/core/server/logging/layouts/json_layout.ts

```diff
@@ -41,8 +43,25 @@ export class JsonLayout implements Layout {
     };
   }
 
+  private static mergeLogMeta(log: Ecs, meta: LogMeta): Ecs {
+    // Paths which cannot be overridden via `meta`.
+    const RESERVED_PATHS = ['ecs', 'message'];
+
+    for (const path of RESERVED_PATHS) {
+      if (has(meta, path)) {
+        // To prevent losing useful data, we "quarantine" conflicting keys
+        // under a `kibana` property rather than deleting them entirely.
+        set(log, `kibana.${path}`, get(meta, path));
+        unset(meta, path);
+      }
```

Comment:
Open for discussion on this.

The point here is to identify any fields which we definitely do _not_ want to be overridden by `LogMeta`. In looking at the existing JSON layout tests, we have ones checking that you can do things like override `@timestamp` and `level`.

TBH I'm not sure what the concrete use cases would be for doing this, but I didn't want to get rid of that functionality without confirming.

So I came up with `ecs` and `message` because:
1. We need to dictate the current ECS version in core (via the new `Ecs` type)
2. Someone who puts a `message` in their `meta` could inadvertently override the main log message.

All this does is removes the offending keys/values, and copies them into a separate `kibana` key. (We can call this key whatever we want, I went with `kibana` initially as it was how audit logging handles custom fields.)

Reply:
Could we enforce these properties aren't overridden using typescript instead of automatically moving them around? I'd be worried that developers will start to rely on that functionality when what we really want is for them not to specify those properties at all. Something along these lines as the log function meta arg could work:

```
type LogMeta = Omit<Ecs, 'ecs' | 'message' | '@timestamp'>
```

Reply:
Yeah that plan makes sense to me and I prefer the simplicity -- so if there's no opposition I will:
- omit ecs, timestamp, message, log level from the types so that folks don't try to provide them
- if they _are_ provided, just ignore it (AFAICT nobody is providing these today in kibana anyway)
- remove the tests that allow folks to override timestamp and log level (assuming we see no use case for it, which I cannot think of)

One of the original reasons for taking the approach above was that someone raised a concern about what happens if someone puts an `Error` object in the log meta, and the error `message` overrides the main `message`, however with the new types in place this shouldn't be possible anyway as TS will complain if you give `LogMeta` an `Error`. (Plus it's the wrong way to go about it: `warn`, `error`, `fatal` all accept `string | Error` in the message).